### PR TITLE
Add rule S6327: Using unencrypted SNS topics is security-sensitive

### LINKED ISSUE #3427: Rule S6327: Using unencrypted SNS topics is security-sensitive (APPSEC-165)
## What
The rule S6327 detects SNS [Topic](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.Topic.html) and [CfnTopic](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.CfnTopic.html) that implicitly do not enable encryption at rest in AWS CDK by not defining a KMS key.
This has been previously implemented for Python. See [SONARPY-1114](https://sonarsource.atlassian.net/browse/SONARPY-1114).

### "Topic" resources
- Fully qualified class name: `aws-cdk-lib.aws_sns.Topic`
- Initializer [`Topic(scope: Construct, id: string, props?: TopicProps)`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.Topic.html#initializer)
- Source: `props` (3rd parameter)
- Property: [`masterKey`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.Topic.html#masterkey)
- Raise an issue if:
  - `props` (3rd parameter) is not provided
  - `props` (3rd parameter) is `undefined`
  - `masterkey` property is not set if `props` is provided
  - `masterkey` property is set to `undefined`

#### Examples

```javascript
    new Topic(this, 'UnencryptedTopic'); // Sensitive: Omitting "masterKey" disables SNS topics encryption. Make sure it is safe here.
```

### "CfnTopic" resources

- Fully qualified class name: `aws-cdk-lib.aws_sns.CfnTopic`
- Initializer [`CfnTopic(scope: Construct, id: string, props?: CfnTopicProps)
`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.CfnTopic.html#initializer)
- Source: `props` (3rd parameter)
- Property: [`kmsMasterKeyId`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sns.CfnTopic.html#kmsmasterkeyid)
- Raise an issue if:
  - `props` (3rd parameter) is not provided
  - `props` (3rd parameter) is `undefined`
  - `kmsMasterKeyId` property is not set if `props` is provided
  - `kmsMasterKeyId` property is set to `undefined`

#### Examples

```javascript
    new CfnTopic(this, 'UnencryptedCfnTopic'); // Sensitive:  Omitting "kmsMasterKeyId" disables SNS topics encryption. Make sure it is safe here.
```

## Code examples

https://github.com/SonarSource/appsec-poc/pull/35

## RSPEC

[RSPEC](https://github.com/SonarSource/rspec/pull/1298).

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #3427

### CODE REVIEW

File: eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      messages: {
+        issue: 'Omitting "{{key}}" disables SNS topics encryption. Make sure it is safe here.',
+      },
+    },
+  },
+);
+
+function checkTopic(key: string) {
+  return (expr: estree.NewExpression, ctx: Rule.RuleContext) => {
+    const props = getValueOfExpression(ctx, expr.arguments[2], 'ObjectExpression');
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:49]: you are not ignoring this case right?

constructor(...foo) DO NOT REPORT

File: eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    function report(node: estree.Node) {
+      ctx.report({
+        messageId: 'issue',
+        data: { key },
+        node,
+      });
+    }
+
+    function getProperty(expr: estree.ObjectExpression, key: string): estree.Property | null {
+      for (const property of expr.properties.reverse()) {
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:76]: array.reverse modifies the AST, what about standard backwards for loop?

**Review by victor-diez-sonarsource — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts (hunk 2 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:76]: Good point.

File: eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts (hunk 1 above)
Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:49]: I forgot. Fixed.

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/linting/eslint/rules/aws-sns-unencrypted-topics.ts:52]: I wonder whether we should move the spreading check in the rule template itself.

@victor-diez-sonarsource WDYT?

**Review by victor-diez-sonarsource — APPROVED** [review]

**Pull request merged**